### PR TITLE
Add Workers Analytics Engine forwarder template

### PR DESCRIPTION
**Pull request by jimhawkridge**
Adding a template for a forwarder that allows writing to Workers Analytics Engine via HTTPS.

### CODE REVIEW

File: worker-analytics-engine-forwarder/src/index.js [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+		indexes: [data.ClientRequestHost || ''], // Supply one index
+		blobs: [
+			// Supply a maximum of 20 blobs (max total data size 5120 bytes)
+			data.RayID || '',
+			data.ClientIP || '',
+			data.ClientRequestMethod || '',
+			data.ClientRequestURI || '',
+			data.EdgeStartTimestamp || '',
+			data.EdgeEndTimestamp || '',
+			data.EdgeResponseStatus || '',
```
Comment (byule) [inline worker-analytics-engine-forwarder/src/index.js:28]: Why not a double? Easier to do comparisons like EdgeResponseStatus > 400 (for example)

File: worker-analytics-engine-forwarder/src/index.js [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+function processLogEntry(ANALYTICS, data) {
+	ANALYTICS.writeDataPoint({
+		indexes: [data.ClientRequestHost || ''], // Supply one index
+		blobs: [
+			// Supply a maximum of 20 blobs (max total data size 5120 bytes)
+			data.RayID || '',
+			data.ClientIP || '',
+			data.ClientRequestMethod || '',
+			data.ClientRequestURI || '',
+			data.EdgeStartTimestamp || '',
```
Comment (byule) [inline worker-analytics-engine-forwarder/src/index.js:26]: Maybe these should also be doubles?  Again makes comparisons easier.

File: worker-analytics-engine-forwarder/wrangler.toml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+name = "workers-analytics-engine-template"
+main = "src/index.js"
+compatibility_date = "2022-10-21"
+
+
+[[unsafe.bindings]]
```
[the remark this answers is not in the document]
Comment (byule) [reply worker-analytics-engine-forwarder/wrangler.toml:6]: +1 - would be nice to add actual WAE support, I imagine it is quite simple.

File: worker-analytics-engine-forwarder/src/index.js [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+	const secret = env.BEARER_TOKEN;
+	if (secret === undefined) {
+		return false;
+	}
+	const want = `Bearer ${secret}`;
+	return authz == want;
+}
+
+// readStream decompresses the incoming stream if needed then slices the stream into lines separated by \n
+async function readStream(body, compressed) {
```
[the remark this answers is not in the document]
Comment (jimhawkridge) [reply worker-analytics-engine-forwarder/src/index.js:87]: Yeah, good thought. I ought to be able to make this async pretty easily.

File: worker-analytics-engine-forwarder/src/index.js [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+// isAuthd checks that the user has supplied an appropriate bearer token in the Authorization header.
+function isAuthd(request, env) {
+	const authz = request.headers.get('Authorization');
+	const secret = env.BEARER_TOKEN;
+	if (secret === undefined) {
+		return false;
+	}
+	const want = `Bearer ${secret}`;
+	return authz == want;
```
Comment (pims) [inline worker-analytics-engine-forwarder/src/index.js:83]: Already discussed internally, but leaving a comment here so we don't forget:
```
const crypto = require('crypto');
[…]
const headerBuffer = Buffer.from(authz);
const secretBuffer = Buffer.from(secret);

return crypto.timingSafeEqual(headerBuffer, secretBuffer);

**Comment by jimhawkridge** [comment]
I've addressed all the PR comments and Wrangler now supports AE bindings without unsafe. So this should be good now.

**Review by codewithkristian — APPROVED** [review]

**Pull request merged**